### DISCUSSION TITLE
Suggestion for a "Line of sight" feature for FPV video / RF links

### DISCUSSION THREAD

**Discussion by mirov** [stronnag/mwptools#157] [Ideas]
I've been working on some missions that twist around some terrain (typically above the launch point) and as you'd expect now and then the plane's path takes it behind some hill that is a bit too high (sometimes pretty far away from any waypoints, but on a line between the plane and the receiver(s) on the ground).  

What I've been doing to visualize this situation is to temporarily add a waypoint at my ground station's location & altitude at the end of the mission, and then add another one that I move around to any suspect areas along the flight path (adjusting manually the altitude for that spot).  Then I run a terrain analysis and see if anything is so high that it blocks or comes close to blocking line-of-sight.  It looks like this approach has been able to find all the spots that actual flying situations have revealed.

Sooo, some sort of LOS analysis tool that let you do a terrain analysis along the path from the launch point to a second point where you've clicked.  There may be a hundred waypoints or interesting spots that you'd want to check, so as much "automagic" behavior here as can be handled would sure make it easier to use (for example once the second point is clicked it would be nice to draw a line on the map in some contrasting color and then update the terrain analysis gnuplot along that path).

**Comment by stronnag** [comment]
In it's simplest form, something like:
* Select  a new option "Point Terrain Analysis" (PTA)
* Drag a distinctive marker (like the "Follow Me" marker)
* Click "OK" in the PTA  dialog
* LOS analysis is performed from planned home to the arbitrary point, line on map in distinctive colour depending on whether LOS or not
* Move the marker, click PTA "OK"  for another point; or
* Click PTA "Close" to finish and remove the LOS lines.

90% of which would reuse existing stuff.

The issue with doing anything more complex, such as a "bird's eye" view with overlaid LOS/noLOS shading is that you really need a local DEM due to number of points required.

So, maybe.

**Comment by stronnag** [comment]
Of course, the interesting it is the elevation required at the clicked point.
* Is the required point projected on to the mission ?
* Does the user input the AGL ?
* Does the user specify 'X' metre along the leg from  WP 'Y' ?

Or am I missing something ?

**Reply by mirov** [reply]
I suppose we can give ourselves some slack here by looking at this as a tool that gives useful results vs absolute precision and absolute specification of parameters.  The selected point's AGL can pretty much be taken as a linear interpolation between the two closest waypoints (there's going to be some fuzziness there).  So MWP would figure out where on the flight path between the two closest WPs it thinks you are indicating and use that <ie: automagic applied>.   

Perhaps you could slide a marker all along the flight path (like a bead on a string) and update the LOS clearance plot whenever you released the mouse button ?

**Reply by mirov** [reply]
Or if you want eye-candy:  use the "preview mission" capability to automatically query all points along the mission, at each point (or every xx meters) run the LOS analysis and <somehow> aggregate them to create a heat map of "shaded" regions where LOS RF is going to be an issue...

**Comment by stronnag** [comment]
There could be a slider which is interpreted as % along the preview path (ignoring such ugliness as multiple or infinite JUMPs). That's probably easier than the geodesy to see where some arbitrary location intercepts the mission.

The second idea is nice, but the online DEMs tend have usage limits, which is the case of mwp as there's a single Bing (free tier) application key across tiles and elevation requests, for all users which I'd prefer not to abuse. I have no means of knowing where on  the usage scale mwp happens to be.

**Reply by mirov** [reply]
Yup, a slider that propels the query point along the mission path is just fine.  A bit manual, but way better than the half dozen clicks and reposition I'm doing for each "manual" probe today.

Yeah, it would really bump up the DEM queries.  Perhaps the overall usage of this feature would be relatively low, can't say.  If you really wanted to complicate your life you could allow for per-user application keys that would each be a tiny fraction of today's aggregate MWP usage on a single key.  And if someone motivated by wanting to maintain video and control around bumpy terrain <cough, Ukranian drone pilots> wanted to buy a key that would be fine.

**Reply by stronnag** [reply]
> Yeah, it would really bump up the DEM queries. Perhaps the overall usage of this feature would be relatively low, can't say. If you really wanted to complicate your life you could allow for per-user application keys that would each be a tiny fraction of today's aggregate MWP usage on a single key. And if someone motivated by wanting to maintain video and control around bumpy terrain <cough, Ukranian drone pilots> wanted to buy a key that would be fine.

I might just force the user to have their own key anyway; after all they can use the same key used for the INAV Configurator.

For now , we'll start with the slider (and the app key) and see how that goes, then maybe add automation **if the user has supplied their own key**.

**Reply by mirov** [reply]
Yeah, this is the sort of thing that you are going to want to play with for a while to see what's useful and what's missing.  It seems like a trivial task to just fly well above the terrain under the plane, but in my situation fairly low hills that are close to my ground station mask reception in a very high shadowed region.

**Comment by stronnag** [comment]
So there will be a manual slider:
![image](https://github.com/stronnag/mwptools/assets/158229/e90281e2-1fbc-4afb-98a8-4fbc25ef056a)
From which "Perform LOS" may be invoked as often as the user requires, resulting in plots, for example:
![s0](https://github.com/stronnag/mwptools/assets/158229/1eb7026f-1ed8-4a18-9c13-773a6bd52bdd)
or
![s1](https://github.com/stronnag/mwptools/assets/158229/7f0b5930-43b3-4077-8cde-4929e8ae9eb9)

**Reply by mirov** [reply]
This is going to be very nice.

**Comment by stronnag** [comment]
Here's where it is now (not yet pushed to GH). My "pathologically interesting" terrain test mission. The red "⨁" icon is the point of interest (POI) which is moved along the mission path by the slider. The slider covers the extent of the mission, with JUMPs reduced to a single iteration for simplicity.

Currently it does not indicate if the POI is within LOS or not, but it's pretty obvious from the plot.

![image](https://github.com/stronnag/mwptools/assets/158229/d93aa4ac-7cfe-4e16-b956-8eef35b3237e)

I'll push a test branch later.

**Comment by mirov** [comment]
Looks nice.  Let me know when there's something to try out.  I've got a
half dozen missions with as many as 60 WP that I've manually scrubbed,
we'll quickly find out if I missed anything.  I think it is fine to not
actually detect the loss of LOS, with trees that can happen anyway even
though you are clear of the terrain.   Much like the Elevation Analysis
tool, the user will want to have some margin (and btw, that margin will be
different for 433/915/2400/5800 MHz links).

-Russ

**Comment by stronnag** [comment]
* Red and green dot dash lines indicating LOS
* Removed when the LOS slider window is closed (as are any stray gnuplot windows)
* Accessed from WP right mouse button --- the slider is initially set to that WP location

![image](https://github.com/stronnag/mwptools/assets/158229/30cc978d-710f-436d-9d15-5265080ec4c2)

You can checkout the `development` branch and give it a try.  As you can see from the image, even on Debian 12 with MateWM.

For now, the user will have to mentally extrapolate any "margin".

**Comment by stronnag** [comment]
If you use **your own** Bing key, then there's a "Auto LOS" option, which iterates around the mission thusly:
![image](https://github.com/stronnag/mwptools/assets/158229/363b6f0e-2533-461c-be09-c9e70dcfb940)

You can set your own bing key in `~/.config/mwp/cmdopts` as for any other environment variable, e.g.:

```
#--debug-flags 20
--dont-maximise

MWP_BLACK_TILE=/home/jrh/.config/mwp/mars.png
MWP_TIME_FMT=%T.%f
MWP_BING_KEY=Axxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

**Comment by mirov** [comment]
Very nice.  I, as well as others, will surely appreciate instructions on
how to get a "Bing Key".

BTW, now and then I find that my home point has been moved while
clicking/panning around a mission.  The one I sent you is such a case, the
home point should be close to WP1 (which changes the LOS results
significantly).

-Russ

**Comment by stronnag** [comment]
> Very nice. I, as well as others, will surely appreciate instructions on how to get a "Bing Key". 

As documented in the [INAV Configurator README](https://github.com/iNavFlight/inav-configurator#how-to-get-the-bing-maps-api-key)

>BTW, now and then I find that my home point has been moved while clicking/panning around a mission. The one I sent you is such a case, the home point should be close to WP1 (which changes the LOS results significantly).

Please provide precise, step by step instructions to reproduce.

**Comment by stronnag** [comment]
With  5238a3ad (which makes auto somewhat smoother), I think we're done for now ... to be left for a few days before promoting to master.

**Comment by stronnag** [comment]
User Guide [preliminary documentation](https://stronnag.github.io/mwptools/mwp-los-tool)

**Comment by mirov** [comment]
Hey there.  Got the latest installed, the new features work great.  Clear
instructions, got my Bing Maps key without any trouble.  Auto-LOS is really
nice, wow what a labor saver !

A few times I've had artifacts left behind on the screen after an Auto-LOS
run:
[image: Screenshot at 2023-11-09 14-19-14.png]

And now and then the Auto-LOS seems to hang.  I'd also earlier seen a hang
when doing "manual" LOS analysis.  Killing MWP and starting again of course
fixes it.  It could easily be Bing not returning data in a timely fashion,
I haven't waited to see if it will eventually timeout.

So let folks know they have a new thing to try & use, it is highly useable
right now.

-Russ

**Comment by stronnag** [comment]
Occasionally I see very short pauses, but never a hang. I believe it's a network glitch getting data from Bing. 
I made a short video. 

https://www.youtube.com/embed/UnejePWHMuM

The video is pretty smooth (recorded on a Wayland display); an earlier X11/Xwayland display had noticeable pauses.

**Comment by stronnag** [comment]
It will be promoted to master and a release built this weekend.

**Comment by stronnag** [comment]
It will be promoted to master and a release built this weekend.

**Comment by mirov** [comment]
It certainly wasn't just a pause.  Specifically I saw it after running
quite a few missions through Auto-LOS as I screened my "library".  I've
also seen the LOS Green/Red lines lag behind the moving marker, normally
near the end of the Auto-LOS run.  It will actually complete and sit there
with the marker stationary at the last WP for ~5-7 seconds and then draw
the stragglers.  It could easily be something on my system, I've never seen
it hang or leave the plots uncompletedd.

Looking forward to the release, and it was sure rewarding to work with you
on this.  You are a very responsive developer.

-Russ

**Comment by mirov** [comment]
Here's a mission that I actually flew last week.  I indeed did have video
fade out between WP6 & WP7-8.  Originally I figured it was just too far,
too many trees.  But run Auto-LOS on it and you'll see there are a couple
of spots in that area that fail.  But pick just about any spot in there and
run a LOS analysis and you'll see that the entire stretch is
super-marginal.

This is a good candidate for some LOS indicator line shading, and for some
"I need at least 10 m of LOS margin" tests.

BTW, the plane flew the path just fine.

-Russ

**Reply by stronnag** [reply]
I think you forgot to add the mission; it would be a useful test case as I'm working on `margin` at the moment.

**Comment by stronnag** [comment]
So margin will be
* Green if LOS with margin
* Yellow (Orange) if LOS above terrain
* Red if no LOG

Additionally

* Margin is scaled along the path between home (otherwise you'd always fail with `-margin` non-LOS at home)
* A "blob" is displayed on non-green LOS lines where LOS is lost.

![image](https://github.com/stronnag/mwptools/assets/158229/2426722b-a342-4b4b-a9a4-3d0bd72b2611)

And you can make multiple runs / switch between point / auto, clear prior tracks etc.

**Comment by mirov** [comment]
I think github tossed the attachment.  I'll send it to you on Discord.

**Comment by stronnag** [comment]
Setting a margin makes things clearer ....

![image](https://github.com/stronnag/mwptools/assets/158229/bb9ec940-5547-4a62-bb88-e92bc879cf1e)

Hope this is what you'r expecting to see

**Comment by mirov** [comment]
Yes, perfect.  And it really does match the 5.8GHz fading that I saw.
Which is comforting, 'cause when I move up the hill a bit to have a higher
point for the ground station it strongly suggests that the new point will
have good video the entire way.

**Comment by stronnag** [comment]
98d086a7 may even be the "final final" ... or not?

**Comment by mirov** [comment]
I'll test it when I get back into my home office in a few more hours...

**Comment by mirov** [comment]
I just installed 98d086a7 (development), not sure if that is older or newer
then the one you mentioned.  Anyway, it seemed to run just fine.  I like
the placement of the "blobs" on top of the point of signal interception.
Margin seems to do the right thing, but it would be nice to run Auto,
twiddle the margin, and then be able to run Auto again without closing and
opening the window (if possible).  Just a shorter iteration loop and makes
it more fluid.  I did find that I could position the slider back at 0% and
it would enable "Auto" again though.

And perhaps running "Auto" should automatically "Clear" at the start.

**Comment by stronnag** [comment]
You can do as many "Auto / change margin" cycles as you wish without closing the window. 

Auto does not set back to 0% by design; if you just have a small orange or red segment, then you can repeatedly cycle over that without having to start from the begining. This saves a lot of time if you have a small NO-LOS segment later in the mission (and you can just "Stop" so you don't have to iterate to the end if you know the"bad" segment is e.g 55-75% of the mission).

* At 0%, only Auto is enabled because "Point" makes no sense.
* At 100%, neither "Point" nor "Auto" is enabled, for the same reason.
* Anywhere else, both are enabled (as long as you pass the `MWP_BING_KEY` test.

And, yes, you ran "final". It will be `master` tomorrow.

**Comment by mirov** [comment]
Oh, and the right-click LOS/Auto-LOS menu choices seem redundant.  I'd just
have "LOS" as the menu choice, but not Automatically run AUTO until the
button is clicked in the popup.

And I think that pressing AUTO should clear the display of earlier results
and then just run it from WP1 rather than somewhere in the middle.  The
button changes to "STOP" while it is running, but it is really more of a
"PAUSE".    But really not much of that really matters, it is so clear
what's happening with the animated marker movement that nobody is going to
be confused.

At this point things are working so well it is probably more valuable (and
fun) to get it out to others for testing.  Do you just let folks stumble
across it, or do you announce on Discord a new feature like this ?  Oh, and
does any of this make it into the INAV Configurator's version of the
Mission Planner ?

**Comment by mirov** [comment]
BTW, earlier in this thread I mentioned that I've been seeing the brown
home point being shifted around on my map unexpectedly (and then being used
as the reference for subsequent LOS stuff).  It looks like a very
repeatable way to make this happen is simply to zoom in so that you can see
a few WPs on the screen, dragging the screen to center a leg.  Turn on
"Edit WPs".   The brown home point icon will show up.  Turn off "Edit WPs",
the home point icon will remain.  Run an Auto-LOS, and you'll see that it
is referencing that spurious home point.

I'm not sure if this is an issue or expected behavior, but it sure is easy
to move the home point and then get all sorts of LOS issues until you
realize things moved.
